File: Sources/Server/Sources/Game/Entity/Dynamics/Mob/MobBodyConnection.ts
import { isPetal } from "../../../../../../Shared/Entity/Dynamics/Mob/Petal/Petal";
import type { WavePool } from "../../../Genres/Wave/WavePool";
import type { EntityMixinConstructor, EntityMixinTemplate } from "../Entity";
import { ON_UPDATE_TICK } from "../Entity";
import type { BaseMob, MobInstance } from "./Mob";
import MOB_PROFILES from "../../../../../../Shared/Native/mob_profiles.json";

const TAU = Math.PI * 2;

/**
 * Get first segment (head) of mob.
 */
export const traverseMobSegments = (poolThis: WavePool, mob: MobInstance): MobInstance => {
    // Walk through segments
    const { connectingSegment } = mob;

    if (connectingSegment && poolThis.getMob(connectingSegment.id)) {
        return traverseMobSegments(poolThis, connectingSegment);
    }

    return mob;
};

/**
 * Determine if mob is segment of body.
 */
export const isBody = (poolThis: WavePool, mob: MobInstance): boolean => traverseMobSegments(poolThis, mob) !== mob;

export function MobBodyConnection<T extends EntityMixinConstructor<BaseMob>>(Base: T) {
    return class extends Base implements EntityMixinTemplate {
        [ON_UPDATE_TICK](poolThis: WavePool): void {
            super[ON_UPDATE_TICK](poolThis);

            // Dont connect when this is petal
            if (isPetal(this.type)) return;

            // If not body, return
            if (!this.connectingSegment) return;

            // If connected segment dead, divide body
            if (!poolThis.getMob(this.connectingSegment.id)) {
                this.connectingSegment = null;

                return;
            }

            const dx = this.connectingSegment.x - this.x;
            const dy = this.connectingSegment.y - this.y;

            const { collision } = MOB_PROFILES[this.type];

            // Arc
            const segmentDistance = (collision.radius * 2) * (this.size / collision.fraction);

            const currentDistance = Math.hypot(dx, dy);

            if (currentDistance > segmentDistance) {
                this.magnitude = 0;
                this.angle = ((Math.atan2(dy, dx) / TAU) * 255 + 255) % 255;

                const ratio = (currentDistance - segmentDistance) / currentDistance;
                this.x += dx * ratio;
                this.y += dy * ratio;
            }
        }
    };
}

File: Sources/goserver/internal/wave/mob_body_connection.go
package wave

import (
	"math"

	"flooooio/internal/native"
)

// traverseMobSegments traverse mob segments and return final segment.
func traverseMobSegments(wp *WavePool, m *Mob) *Mob {
	if m.ConnectingSegment != nil {
		if segment, ok := m.ConnectingSegment.(*Mob); ok && !segment.WasEliminated(wp) {
			return traverseMobSegments(wp, segment)
		}
	}

	return m
}

// isBody determinate if segment piece mob is body.
func isBody(wp *WavePool, m *Mob) bool {
	return traverseMobSegments(wp, m) != m
}

func (m *Mob) MobBodyConnection(wp *WavePool) {
	if m.ConnectingSegment == nil {
		return
	}

	if IsDeadNode(wp, m.ConnectingSegment) {
		m.ConnectingSegment = nil

		return
	}

	dx := m.ConnectingSegment.GetX() - m.X
	dy := m.ConnectingSegment.GetY() - m.Y

	mc := native.MobProfiles[m.Type].Collision

	// Arc
	segmentDistance := (mc.Radius * 2) * (m.Size / mc.Fraction)

	currentDistance := math.Hypot(dx, dy)

	if currentDistance > segmentDistance {
		m.Magnitude = 0
		m.Angle = math.Mod((math.Atan2(dy, dx)/tau)*255+255, 255)

		ratio := (currentDistance - segmentDistance) / currentDistance
		m.X += dx * ratio
		m.Y += dy * ratio
	}
}
